src/query_graph/graph_path.rs

```diff
@@ -0,0 +1,228 @@
+///
+/// Additionally, for each edge of the path, a `GraphPath` records the "trigger" that made the
+/// traversal take that edge. In practice, the "trigger" can be seen as a way to decorate a path
+/// with some additional metadata for each element of the path. In practice, that trigger is used in
+/// 2 main ways (corresponding to our 2 main query graph traversals):
+/// - For composition validation, the traversal of the federated query graph is driven by other
+///   transitions into the supergraph API query graph (essentially, composition validation is about
+///   finding, for every path in supergraph API query graph, a "matching" traversal of the federated
+///   query graph). In that case, for the graph paths we build on the federated query graph, the
+///   "trigger" will be one of the edge transitions from the supergraph API query graph (which,
+///   granted, will be fairly similar to the one of the edge we're taking in the federated query
+///   graph; in practice, triggers are more useful in the query planning case).
+/// - For query planning, the traversal of the federated query graph is driven by the elements of
+///   the query we are planning. Which means that the "trigger" for taking an edge in this case will
+///   be an operation element (or `None`). See the specialized `OpGraphPath` that is defined for this
+///   use case.
+///
+/// Lastly, some `GraphPath`s can actually encode `None` edges: this is used during query planning
+/// in the (rare) case where the query we plan for has an inline fragment spread without type
+/// condition (or a "useless" one, i.e. one that doesn't restrict the possible types anymore than
+/// they already were) but with some directives. In that case, we want to preserve the information
+/// about the directive (to properly rebuild query plans later) but it doesn't correspond to taking
+/// any edges, so we add a `None` edge and use the trigger to store the fragment spread.
+///
+/// Regarding type parameters:
+/// - `TTrigger`: The type of the path's "triggers", metadata that can associated to each element
+///   of the path (see above for more details).
+/// - `TEdge`: The type of the edge. Either `Option<EdgeIndex>` (meaning that the path may have a
+///   `None` edge) or `never` (the path cannot have `None` edges).
+// PORT_NOTE: The JS codebase also parameterized whether the head of the path was a root node, but
+// in the Rust code we don't have a distinguished type for that case. We instead check this at
+// runtime (introducing the new field `head_must_be_root`). This means the `RootPath` type in the
+// JS codebase is replaced with this one.
+#[derive(Debug, Clone)]
+pub(crate) struct GraphPath<TTrigger: Eq + Hash, TEdge: Into<Option<EdgeIndex>>> {
+    /// The query graph of which this is a path.
+    graph: Arc<QueryGraph>,
+    /// The node at which the path starts (the head node of the first edge in path, aliased here for
+    /// convenience).
+    head: NodeIndex,
```

Comment:
Rather than have duplicated data that could accidentally get out of sync, would a method provide the same convenience? This same info is in `self.edges.first()`, right?

(Same for `tail` and `edge_to_tail`)

Reply:
I think the `head`/`tail` actually point to the first/last non-null edge in the list (and if no such edges exist, `head`/`tail` point to the same node). I've updated the doc strings for them to be more clear at least. I think we can eliminate `edge_to_tail` (in the JS codebase it was eliminating two layers of indirection while here in Rust it's only eliminating one). Similarly, I'll also remove `SubgraphEnteringEdgeInfo.edge`.